tianshou/policy/modelfree/bcq.py

```diff
@@ -0,0 +1,138 @@
+
+        def forward(self, state):
+            q = F.relu(self.q1(state))
+            q = F.relu(self.q2(q))
+
+            i = F.relu(self.i1(state))
+            i = F.relu(self.i2(i))
+            i = F.relu(self.i3(i))
+            return self.q3(q), F.log_softmax(i, dim=1), i
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        tau: float,
+        target_update_frequency: int,
+        device: str,
+        gamma: float,
+        imitation_logits_penalty: float,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self._policy_net = model
+        self._optimizer = optim
+        self._cnt = 0
+        self._device = device
+        self._gamma = gamma
+        self._tau = tau
+        self._target_net = deepcopy(self._policy_net)
+        self._target_net.eval()
+        self._target_update_frequency = target_update_frequency
+        self._imitation_logits_penalty = imitation_logits_penalty
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        **kwargs: Any,
+    ) -> Batch:
+        batch.to_torch()
+
+        state = batch.obs
+        q, imt, _ = self._policy_net(state.float())
+        imt = imt.exp()
+        imt = (imt / imt.max(1, keepdim=True)[0] > self._tau).float()
+
+        # Use large negative number to mask actions from argmax
+        action = (imt * q + (1.0 - imt) * -1e8).argmax(1)
+
+        return Batch(act=action, state=state, qvalue=q)
+
+    def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
+        batch.to_torch()
+
+        non_final_mask = torch.tensor(
+            tuple(map(lambda s: not s, batch.done)),
+            device=self._device,
+            dtype=torch.bool,
+        )
+
+        try:
+            non_final_next_states = torch.cat(
+                [s.obs.unsqueeze(0) for s in batch if not s.done], dim=0
+            )
+        except Exception:
+            non_final_next_states = None
+
+        # Compute the target Q value
+        with torch.no_grad():
+            expected_state_action_values = batch.rew.float()
+
+            # Add target Q value for non-final next_state
+            if non_final_next_states is not None:
+                q, imt, _ = self._policy_net(non_final_next_states)
+                imt = imt.exp()
+                imt = (imt / imt.max(1, keepdim=True)[0] > self._tau).float()
+
+                # Use large negative number to mask actions from argmax
+                next_action = (imt * q + (1 - imt) * -1e8).argmax(1, keepdim=True)
+
+                q, _, _ = self._target_net(non_final_next_states)
+                q = q.gather(1, next_action).reshape(-1, 1)
+
+                next_state_values = torch.zeros(len(batch), device=self._device).float()
+                next_state_values[non_final_mask] = q.squeeze()
```

Comment:
Current script cannot run on cuda device.
```
Traceback (most recent call last):
  File "test/discrete/test_bcq.py", line 100, in <module>
    test_bcq(get_args())
  File "test/discrete/test_bcq.py", line 94, in test_bcq
    args.test_frequency,
  File "/home/trinkle/github/tianshou-new/tianshou/trainer/offline.py", line 30, in offline_trainer
    loss = policy.update(batch_size, buffer)
  File "/home/trinkle/github/tianshou-new/tianshou/policy/base.py", line 171, in update
    result = self.learn(batch, **kwargs)
  File "/home/trinkle/github/tianshou-new/tianshou/policy/modelfree/bcq.py", line 123, in learn
    next_state_values[non_final_mask] = q.squeeze()
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```

Reply:
Will have it tested on a GPU machine.